appengine/flexible/TwelveFactor/Services/CloudStorageFileProvider.cs

```diff
@@ -0,0 +1,259 @@
+                    path.ObjectName);
+                return null;
+            }
+        }
+
+        /// <summary> Watch a file for changes.
+        public IChangeToken Watch(string filter)
+        {
+            if (_pollingInterval == null) {
+                return null;
+            }
+            _logger.LogDebug("Watching {0}", filter);
+            var path = SplitObjectPath(filter);
+            var token = new CloudStorageChangeToken();
+            var obj = GetObject(path);
+            long? generation = obj?.Generation;
+            Task.Run(async () => {
+                while (true) {
+                    await Task.Delay(_pollingInterval.Value);
+                    try
+                    {
+                        obj = GetObject(path);
+                        if (generation != obj?.Generation) {
+                            token.HasChanged = true;
+                            return;
+                        }
+                    } 
+                    catch (Exception e) 
+                    {
+                        _logger.LogError(0, e, "Exception while watching {0}",
+                            filter);
+                    }
+                }
+            });
+            return token;            
+        }            
+    }
+
+    class CloudStorageChangeToken : IChangeToken
+    {
+        List<ChangeCallback> _callbacks = new List<ChangeCallback>();
+        bool _hasChanged = false;
+
+        public bool HasChanged { 
+            get { return _hasChanged; }
+            set {
+                _hasChanged = true;
+                foreach (var callback in _callbacks) {
+                    callback.Invoke();
+                }
+            }
+        }
+
+        public bool ActiveChangeCallbacks => true;
+
+        public IDisposable RegisterChangeCallback(Action<object> callback,
+            object state)
+        {
+            var disposable = new ChangeCallback() {
+                Callback = callback,
+                State = state
+            };
+            _callbacks.Add(disposable);
+            return disposable;
+        }
+
+        class ChangeCallback : IDisposable {
+            public Action<object> Callback { get; set; }
+            public object State {get; set;}
+
+            private object _lock = new object();
+            private bool _done = false;
+            public void Dispose()
+            {
+                lock(_lock) {
+                    if (!_done) {
+                        _done = true;
+                    }
+                }
+            }
+
+            public void Invoke() {
+                lock(_lock) {
+                    if (!_done) {
+                        _done = true;
+                        Callback(State);
+                    }
+                }                
+            }
+        };
+    }
+
+    class CloudStorageFileInfo : IFileInfo
+    {
+        readonly Google.Apis.Storage.v1.Data.Object _cloudStorageObject;

```

Comment:
Nit, you could use a name alias to make this more readable:
```C#
using GcsObject = Google.Apis.Storage.v1.Data.Object;
```

Then you could just refer to `GcsObject` in your code.